### ISSUE TITLE
Navigation of top window from within iframe doesn't work in IOS

### ISSUE THREAD

**Issue by omerr132** [ampproject/amphtml#34934]
### Description

When viewing an AMP page from the Google search results on IOS, navigating the top window from within iframes does not work, as if the 'allow-top-navigation' permission is not set (and it is).
This only happens in IOS, when viewing AMP pages from the Google search results app (not via the browser). 
Tested on several iphones. 

### Reproduction Steps

1. Use an iphone, and open the Google Search app (not via the browser)
2. search the term (apologies for the language) "אשת מלונות בארץ"
3. open the top most AMP result.
4. In the opened page, make some selections. [See screenshot for instructions](https://user-images.githubusercontent.com/25805938/122719492-ea12b300-d276-11eb-8ed0-72f0d2d96756.png).
5. Click the blue button.

**Result**: Nothing happens.
**Expected result** the top page should have been redirected to a different URL (see how the same page behaves on other browsers: https://www.google.com/amp/s/www.eshet.com/domestichotels/amp/)

Video of the issue: https://drive.google.com/file/d/1M_iQkQVLm_4I1mZN4oY4UR4u6hWWAsu0/view?usp=sharing

### OS(s) Affected

IOS15

### Device(s) Affected

All iphones

**Comment by micajuine-ho** [comment]
@omerr132 When I try to load the page I get a 403 error, for the section of the page where I'm supposed to make the selections (I am in US). 

> navigating the top window from within iframes does not work

If I'm understanding correctly, clicking the blue button should remove the iframe? 

What mechanism are you using within the iframe to try to remove itself?

**Comment by omerr132** [comment]
@micajuine-ho  Clicking on the blue button doesn't remove the iframe, it redirects the top window. You can see it in play here: https://www.google.com/amp/s/www.eshet.com/domestichotels/amp/. This is the exact same page,  only when you view it not in the context of IOS AMP viewer - it works.

**Comment by micajuine-ho** [comment]
> You can see it in play here: https://www.google.com/amp/s/www.eshet.com/domestichotels/amp/. 

![image](https://user-images.githubusercontent.com/30711678/122789093-cbf38580-d26b-11eb-9239-1edeacf9b10b.png)

I cannot see it.

**Comment by omerr132** [comment]
I'll get them to open it. 

As for your 2nd question - this is what's taking place when the blue button is clicked.  

window.open(`${pathname}?${stringify(searchQuery)}`, "_top")

**Comment by micajuine-ho** [comment]
> window.open(${pathname}?${stringify(searchQuery)}, "_top")

cc @ampproject/wg-components  @ampproject/wg-performance Does anyone have any ideas as to why this wouldn't work in the google results app?

**Comment by omerr132** [comment]
+ it happens only on IOS. Here's a video of it working as expected in Android chrome https://drive.google.com/file/d/1Q28qXSVHjfDAlTyL8Ax9quXEWdhfxj8T/view?usp=sharing

**Comment by alanorozco** [comment]
This fails on the Viewer since the Google-level iframe that holds the AMP document lacks `allow-top-navigation`. (The `<amp-iframe>` has it set, but it has to be set throughout the whole nesting chain.)

It's a security concern why the top-level iframe cannot have the `allow-top-navigation` sandbox setting, so I'll defer to @ampproject/wg-security-privacy for that question.

@omerr132 Have you considered using something other than an `<iframe>` to render the form? Maybe `<amp-script>`?

**Comment by omerr132** [comment]
@alanorozco  - 
+ Is this new? We encountered this issue for the first time last week even though the site has been live for longer.
+ As for amp-script, any solution that can't be implemented with an iframe will be a deal breaker unfortunately. 
+ iframes are a versatile solution for many AMP sites, especially those who weren't built straight into AMP. Blocking this use case is a serious limitation.

**Comment by honeybadgerdontcare** [comment]
@molnarg could you please address the question of why the top-level iframe cannot have the `allow-top-navigation` sandbox setting?

**Comment by omerr132** [comment]
How do AMP elements which redirect into new pages, like Facebook, Twitter, etc. behave under this restriction?

**Comment by alanorozco** [comment]
> How do AMP elements which redirect into new pages, like Facebook, Twitter, etc. behave under this restriction?

@omerr132 I'd like to verify this difference by debugging your document.
Could you please let us know once the document inside the `<amp-iframe>` is accessible?

**Comment by omerr132** [comment]
@alanorozco I will. Hopefully by tomorrow. I suspect it's an IP thing - so in the meanwhile please try to use a proxy.
Until then, I'd like to stress the significance of this restriction - even though iframe is not the optimal way to go about AMP, it is in fact used a lot, not only by us. This will break a whole lot of AMP sites, if it hadn't already.

**Comment by alanorozco** [comment]
@omerr132 I agree that this would ideally just work. Please be patient, I'm trying to understand the details.

I can't make an assessment without verifying the problem on a working page. I don't think it's advisable for me to configure a VPN on my personal iOS device in order to look at a single document, which I hope you understand.

Looking forward to debugging more once the document is accessible.

**Comment by omerr132** [comment]
@alanorozco This page shouldn't have access limitation:
https://convert.ampify.io/5f12e587a2a85c002884bfe5?u=https://www.eshet.com/domestichotels/

**Comment by omerr132** [comment]
Any update on the status of this issue?

**Comment by alanorozco** [comment]
@omerr132 I can't open the `convert.ampify.io` link from the viewer, since it's not cached. 

When I open your previous link, it's stuck on a redirect loop.

**Comment by omerr132** [comment]
@molnarg thanks for your reply. 

> Top level navigation should be restricted in situations when the AMP page is displayed in a WebView that has a custom UI around it which assumes that the content rendered cannot navigate away by itself.

From my understanding, viewing AMP in the WebView of the Google Search App on both Android and IOS fits the described situation (as there's no URL bar). Google does support the option to navigate away from the AMP page while still in the WebView (and it updates the WebView UI to display the URL/domain of the new content). However if the top navigation stems from within an iframe inside the AMP, it is blocked. If the UI manages the navigation, why this specific navigation shouldn't work?

> If this is some custom AMP rendering WebView that is triggered when clicking on an AMP search result, then it could be a concern. These custom AMP rendering WebView UIs should override the top level navigation functions so that they can properly open the clicked links in the default browser or Chrome Custom Tab etc., instead of navigating away the WebView.

Are you referring to the AMP viewer of the Google Search app (since you mentioned search result)? Either way I guess this is a decision that is being made by the owners of the custom WebView, i.e. they can decide to permit any form of navigation, right?
